Жестков А. А. Лаб. 4 Группа 6313

mappy12 opened the pull request:
**ФИО:** Жестков Андрей
**Номер группы:** 6313  
**Номер лабораторной:** 4
**Краткое описание исходного кода:** Исходный код был предоставлен преподавателем
**Краткое описание проделанной работы:** Я выполнил рефакторинг предоставленного кода.

Darktezi commented on lab_2/Final.py:
```diff
@@ -0,0 +1,319 @@
+    return dataframeN
+def numeric(dataframe) :
+    # Numerise bank-full
+    dataframeN = dataframe
+    job_mapping = {'student':0,'unemployed':1, 'unknown':2, 'housemaid':3, 'blue-collar':4, 'services':5, 'retired':6, 'self-employed':7, 'technician':8 ,'management':9,'entrepreneur':10,'admin.':11 }
+    dataframeN = dataframeN.assign(jobN = dataframe['job'].map(job_mapping))
```
Необходимо исправить длинну строк согласно PEP-8

alxmcs commented on lab_2/Final.py:
```diff
@@ -0,0 +1,319 @@
+        dataframe.drop(columns=['month'],inplace= True)
+    return dataframe
+
+def numerique(dataframe) :
+    # Numerise student-por
+    dataframeN = dataframe
```
А долго ли оперативка сможет вывозить создание глубокой копии датафрейма на каждый вызов функции?

Darktezi replied:
функция вызывается один раз в maindeux(аналог main), поэтому я думаю что проблем с памятью быть не должно. Если функция maindeux будет вызываться многократно, то можно попробовать передать копию заранее, если это возможно:
`df_studentN = numerique(df_student.copy())`

alxmcs commented on lab_2/Final.py:
Хотелось бы получить некоторый комментарий насчет того, стоит ли оставлять весь имеющийся тут функционал - манипуляции датафреймом,  обучение и инференс классификаторов, демонстрацию графиков - в единственном модуле?

Darktezi replied:
Для улучшении читаемости кода, тестируемости и поддерживаемости кода необходимо разделить код на несколько модулей. Например:
1. Модуль обработки данных
2. Модуль обучения моделей
3. Модуль оценки и метрик
4. Модуль визуализации

alxmcs commented on lab_2/Final.py:
```diff
@@ -0,0 +1,319 @@
+
+    # Logistic Regression
+    regression_model = LogisticRegression()
+    regression_model.fit(x_train_scaled, y_train)
+    y_pred_regression = regression_model.predict(x_test_scaled)
+    df_Reg = pd.DataFrame(x_test)
+    df_Reg['yN'] = y_pred_regression
+    metric_reg = metric_privilege(df_Reg)
+
+    # KNN
+    knn_model = KNeighborsClassifier(n_neighbors=20)
+    knn_model.fit(x_train_scaled, y_train)
+    y_pred_knn = knn_model.predict(x_test_scaled)
+    df_knn = pd.DataFrame(x_test)
+    df_knn['yN'] = y_pred_knn
+    metric_knn = metric_privilege(df_knn)
+
+    # Random Forest
+    rf_model = RandomForestClassifier(n_estimators=100)
+    rf_model.fit(x_train_scaled, y_train)
+    y_pred_rf=rf_model.predict(x_test_scaled)
+    df_rf = pd.DataFrame(x_test)
+    df_rf['yN'] = y_pred_rf
+    metric_rf = metric_privilege(df_rf)
+
+    # Neural Network
+    nn_model = MLPClassifier(hidden_layer_sizes=(100,), max_iter=500, random_state=1)
+    nn_model.fit(x_train_scaled, y_train)
+    y_pred_nn = nn_model.predict(x_test_scaled)
+    df_nn = pd.DataFrame(x_test)
+    df_nn['yN'] = y_pred_nn
+    metric_nn = metric_privilege(df_nn)
+
+    # SVM
+    svm_model = SVC()
+    svm_model.fit(x_train, y_train)
+    y_pred_svm = svm_model.predict(x_test)
+    df_svm = pd.DataFrame(x_test)
+    df_svm['yN'] = y_pred_svm
+    metric_svm = metric_privilege(df_svm)
```
Этот код выглядит подозрительно подходящим для того, чтобы параметризовать выбор классификатора. Хотелось бы увидеть небольшой пример кода, как это можно сделать

Darktezi replied:
Создаём словарь моделей с указанием на необходимость масштабирования. Функция станет короче и более гибкой, а так же необходимо сделать более понятное название для неё:
```
def train_and_evaluate_model(x_train, x_test, y_train, y_test):
    global df_original
    df_original = pd.DataFrame(x_test)
    df_original['yN'] = y_test

    scaler = StandardScaler()
    x_train_scaled = scaler.fit_transform(x_train)
    x_test_scaled = scaler.transform(x_test)

    classifiers = {
        "Naive Bayes": (MultinomialNB(), False),
        "Logistic Regression": (LogisticRegression(), True),
        "KNN": (KNeighborsClassifier(n_neighbors=20), True),
        "Random Forest": (RandomForestClassifier(n_estimators=100), True),
        "Neural Network": (MLPClassifier(hidden_layer_sizes=(100,), max_iter=500, random_state=1), True),
        "SVM": (SVC(), False)
    }

    metrics_results = []

    for name, (model, use_scaling) in classifiers.items():
        x_train_data, x_test_data = (x_train_scaled, x_test_scaled) if use_scaling else (x_train, x_test)

        model.fit(x_train_data, y_train)
        y_pred = model.predict(x_test_data)

        df_result = pd.DataFrame(x_test)
        df_result['yN'] = y_pred

        metric = metric_privilege(df_result)
        metrics_results.append(metric)

    return metrics_results
```

alxmcs reviewed (approved):
Зачтено

mappy12 committed:
добавлен исходный код

mappy12 committed:
переписан класс Entry, который теперь использует декоратор dataclass

mappy12 committed:
Вместо импорта всего модуля теперь импортируется лишь нужный класс. Улучшена читаемость + очевидно что именно было импортировано

mappy12 committed:
вместо предыдущего кода использована конструкция with open(), тк при предыдущем способе файл открывается, но не закрывается, что нарушает принцип управления ресурсами

mappy12 committed:
добавлен typehint параметра

mappy12 committed:
имя метода getData() изменено на get_data(), тк основная масса методов написана с использованием snake_case

mappy12 committed:
еще одно имя метода в snake_case

mappy12 committed:
entry - это объект класса, а не словарь, поэтому обращаться к полям необходимо через точку.

Также в классе Entry поле "passport_name" определено как строка, но метод "check_passport" ожидает почему то int, что неверно. исправлено

mappy12 committed:
однотипные методы (check_.....) заменены на универсальную функцию валидации

все правила валидации собраны в словаре rules

mappy12 committed:
добавлен метод copy в класс Entry

mappy12 committed:
добавлена конструкция with open

mappy12 committed:
лучше все выполнять внутри функций

mappy12 committed:
~

IlyaOv reviewed (approved):
Зачтено
